src/menu/index.tsx

```diff
@@ -0,0 +1,267 @@
+	options: MenuOption[];
+	/** The initial selected value */
+	initialValue?: string;
+	/** Callback called when user selects a value */
+	onValue(value: string): void;
+	/** Called to request that the menu be closed */
+	onRequestClose?(): void;
+	/** Optional callback, when passed, the widget will no longer control it's own active index / keyboard navigation */
+	onActiveIndexChange?(index: number): void;
+	/** Optional proprty to set the activeIndex when it is being controlled externally */
+	activeIndex?: number;
+	/** Determines if the widget can be focused or not. If the active index is controlled from elsewhere you may wish to stop the menu being focused and receiving keyboard events */
+	focusable?: boolean;
+	/** Callback called when menu root is focused */
+	onFocus?(): void;
+	/** Callback called when menu root is blurred */
+	onBlur?(): void;
+	/** Property to determine how many items to render. Not passing a number will render all results */
+	itemsInView?: number;
+	/** Custom renderer for item contents */
+	itemRenderer?(properties: ItemRendererProperties): RenderResult;
+}
+
+export interface ItemRendererProperties {
+	value: string;
+	label?: string;
+	disabled: boolean;
+	active: boolean;
+	selected: boolean;
+}
+
+interface MenuICache {
+	value: string;
+	initial: string;
+	activeIndex: number;
+	itemsInView: number;
+	menuHeight: number;
+	itemHeight: number;
+	itemToScroll: number;
+	resetInputTextTimer: NodeJS.Timer;
+	inputText: string;
+}
+
+const offscreenHeight = (dnode: RenderResult) => {
+	const r = renderer(() => dnode);
+	const div = global.document.createElement('div');
+	div.style.position = 'absolute';
+	global.document.body.appendChild(div);
+	r.mount({ domNode: div, sync: true });
+	const dimensions = div.getBoundingClientRect();
+	global.document.body.removeChild(div);
+	return dimensions.height;
+};
+
+const factory = create({
+	icache: createICacheMiddleware<MenuICache>(),
+	focus,
+	dimensions,
+	theme
+}).properties<MenuProperties>();
+
+export const Menu = factory(function({
+	properties,
+	middleware: { icache, focus, dimensions, theme }
+}) {
+	const {
+		options,
+		initialValue,
+		onValue,
+		onRequestClose,
+		onActiveIndexChange,
+		activeIndex,
+		focusable = true,
+		onBlur,
+		onFocus,
+		itemsInView,
+		itemRenderer,
+		theme: themeProp
+	} = properties();
+
+	if (initialValue !== undefined && initialValue !== icache.get('initial')) {
+		icache.set('initial', initialValue);
+		icache.set('value', initialValue);
+		icache.set('activeIndex', findIndex(options, (option) => option.value === initialValue));
+	}
+
+	if (itemsInView && itemsInView !== icache.get('itemsInView')) {
+		icache.set('itemsInView', itemsInView);
+
+		const offscreenItemProps = {
+			selected: false,
+			onSelect: () => {},
+			active: false,
+			onRequestActive: () => {},
+			onActive: () => {},
+			scrollIntoView: false
+		};
+
+		const itemHeight = icache.getOrSet(
+			'itemHeight',
+			offscreenHeight(
```

Comment:
I do think these things are clearer when they are not inlined.

Reply:
i thought i already did that... looking